src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,111 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getFirstName)
+                .sorted(Collections.reverseOrder())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted(Comparator.comparing(User::getAge).reversed()
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getPrivileges)
+                .flatMap(List::stream)
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getAge() > age &&
+                        user.getPrivileges().contains(Privilege.UPDATE))
+                .findAny();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users
+                .stream()
+                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToInt(User::getAge)
+                .average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(User::getLastName, Collectors.counting()))
+                .entrySet()
+                .stream()
+                .collect(Collectors.groupingBy(Map.Entry::getValue,
+                        Collectors.mapping(Map.Entry::getKey, Collectors.toList())))
+                .entrySet()
+                .stream()
+                .max(Map.Entry.comparingByKey())
+                .map(Map.Entry::getValue)
+                .filter(x -> x.size() == 1)
+                .map(List::stream)
+                .flatMap(Stream::findAny);
     }
 
+    @SafeVarargs
```

Comment:
Oh nice. Can you say in which situations and why we should use the safevarargs?

Reply:
The @SafeVarargs annotation is used to suppress warnings about unsafe operations at compile time. unsafe operations appear whenever we use varargs. But at the same time, the method must be either final or static